File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    String r = "";
    Boolean[][] memo = null;
    String s;
    
    public String longestPalindrome(String s) {
        int n = s.length();
        memo = new Boolean[n][n];
        this.s = s;
        
        for(int i = 0; i < n; i++) 
            for(int j = 0; j < n; j++)
                helper(i, j);
        
        return r;
    }
    
    boolean helper(int i, int j) {
        if (i > j) return true;
        
        if (memo[i][j] != null) {
            return memo[i][j];
        }
        
        boolean res = (s.charAt(i) == s.charAt(j) && helper(i + 1, j - 1));
        memo[i][j] = res;
        
        if (res && (j - i + 1) > r.length()) { 
            r = s.substring(i, j + 1);
        }
        
        return res;
    }
}

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.scala
object Solution {
  def longestPalindrome(s: String): String = {
    var ml = 1
    var res = s.substring(0, 1)

    val dp: Array[Array[Boolean]] = Array.fill(s.length + 1) {
      Array.fill(s.length)(false)
    }

    for (i <- 0 to s.length - 1) {
      for (j <- 0 to s.length - 1) {
        dp(i)(j) =
          (j - i >= 0 && (i < 2 || dp(i - 2)(j - 1)) && s.charAt(j) == s.charAt(j - i))
        if (dp(i)(j) && i >= ml) {
          ml = i + 1
          res = s.substring(j - i, j + 1)
        }
      }
    }

    res
  }
}